File: Problem_1_recursive.cpp
//Time Complexity: 2^n, where n is the size of the array.
//Space Complexity: O(1) 
class Solution {
private:
     int helper(int comp[], int index, int sum, int size){
        if(index >= size){
            return sum;
        }
        int case_1 = helper(comp, index+1, sum, size);
        int case_2 = helper(comp, index+2, sum+comp[index], size);
        return max(case_1,case_2);
    }
    
    
public:
    int deleteAndEarn(vector<int>& nums) {
        if(nums.size()==0){
            return 0;
        }
        int max = 0;
        for(int i=0; i<nums.size(); i++){
            max = std::max(max,nums.at(i));
        }
        int comp[max+1];
        for(int i=0; i<max+1; i++){
            comp[i] = 0;
        }
        for(int i=0; i<nums.size(); i++){
            comp[nums.at(i)] += nums.at(i);
        }
        
        return helper(comp, 0, 0, max+1);
    }
   
};

File: Problem1_Recursive.java
/**
LeetCode Submitted : NA (Recursion Method)
Time Complexity : O(2^N)
Space Complexity : O(2^N)
Concept : To look for decision to to either choose current problem or not choose current problem.
Any issues : I think the main issue sfor me here was to handle cloning of the array which took some time to figure out
**/


class Solution {
    public int deleteAndEarn(int[] nums) {
        int max = 0;
        
        return Math.max(decision(nums,0,max),decision(nums,1,max));
        
    }
    
    public int decision(int[] nums, int i, int max){
        
        int[] temp = nums.clone();
        int temp_max = max;
        
        
        while(i < nums.length && nums[i] == 0){
            i++;
        }
        
        if(i == nums.length)
            return max;
        
        for(int j= i; j< nums.length;j++){
            if((nums[j] == nums[i] + 1) || (nums[j] == nums[i] - 1)){
                max = max + nums[j];
                nums[j] = 0;
            }
        }
        
        return (Math.max(decision(nums,i+1,max),decision(temp,i+1,temp_max)));
